File: Other/R/data_manager.R
# MSI Analysis Application - Data Manager
# データ管理モジュール

#' データフォルダ内のMSIファイル一覧を取得
#' @param data_folder データフォルダパス
#' @return ファイル名のベクトル（拡張子なし）
list_msi_files <- function(data_folder) {
  if (!dir.exists(data_folder)) {
    return(character())
  }
  
  files <- list.files(data_folder, pattern = "\\.txt$", full.names = FALSE)
  # 拡張子を除去
  tools::file_path_sans_ext(files)
}

#' MSIファイルの妥当性チェック
#' @param file_path ファイルパス
#' @return TRUE/FALSE
validate_msi_file <- function(file_path) {
  if (!file.exists(file_path)) {
    return(list(valid = FALSE, message = "ファイルが見つかりません"))
  }
  
  tryCatch({
    lines <- readLines(file_path, n = 5, warn = FALSE)
    if (length(lines) < 5) {
      return(list(valid = FALSE, message = "ファイルの行数が不足しています (最低5行必要)"))
    }
    
    # 4行目以降がタブ区切りデータかチェック
    data_line <- strsplit(lines[5], "\t")[[1]]
    if (length(data_line) < 4) {
      return(list(valid = FALSE, message = "データ形式が正しくありません"))
    }
    
    return(list(valid = TRUE, message = "OK"))
  }, error = function(e) {
    return(list(valid = FALSE, message = paste("読み込みエラー:", e$message)))
  })
}

#' MRMファイルの読み込み
#' @param mrm_path MRMファイルパス（xlsx/csv）
#' @return MRMデータフレームまたはNULL
load_mrm_file <- function(mrm_path) {
  if (is.null(mrm_path) || mrm_path == "" || !file.exists(mrm_path)) {
    return(NULL)
  }
  
  ext <- tolower(tools::file_ext(mrm_path))
  
  tryCatch({
    if (ext %in% c("xlsx", "xls")) {
      if (requireNamespace("readxl", quietly = TRUE)) {
        df <- as.data.frame(readxl::read_excel(mrm_path))
      } else {
        warning("readxlパッケージがインストールされていません")
        return(NULL)
      }
    } else {
      df <- read.csv(mrm_path, stringsAsFactors = FALSE, check.names = FALSE)
    }
    
    if (nrow(df) == 0) return(NULL)
    
    return(df)
  }, error = function(e) {
    warning("MRMファイル読み込みエラー: ", e$message)
    return(NULL)
  })
}

#' 出力結果フォルダの一覧を取得
#' @param base_dir ベースディレクトリ
#' @return 結果フォルダ情報のデータフレーム
list_result_folders <- function(base_dir) {
  if (!dir.exists(base_dir)) {
    return(data.frame(
      name = character(),
      path = character(),
      date = character(),
      stringsAsFactors = FALSE
    ))
  }
  
  # フォルダ名パターン: *_YYYYMMDD または *_YYYYMMDD_Re-Analysis
  dirs <- list.dirs(base_dir, recursive = FALSE, full.names = TRUE)
  
  result_dirs <- lapply(dirs, function(d) {
    name <- basename(d)
    # RDS_Filesサブフォルダがあれば解析結果フォルダと判定
    if (dir.exists(file.path(d, "RDS_Files")) || 
        length(list.files(d, pattern = "\\.(png|csv)$")) > 0) {
      # 日付部分を抽出 (末尾の8桁数字)
      date_match <- regmatches(name, regexpr("\\d{8}", name))
      date_str <- if (length(date_match) > 0) {
        paste0(substr(date_match, 1, 4), "/", substr(date_match, 5, 6), "/", substr(date_match, 7, 8))
      } else {
        "Unknown"
      }
      return(data.frame(name = name, path = d, date = date_str, stringsAsFactors = FALSE))
    }
    return(NULL)
  })
  
  result_dirs <- result_dirs[!sapply(result_dirs, is.null)]
  if (length(result_dirs) == 0) {
    return(data.frame(name = character(), path = character(), date = character(), stringsAsFactors = FALSE))
  }
  
  do.call(rbind, result_dirs)
}

#' 結果フォルダ内の画像ファイル一覧を取得
#' @param result_dir 結果フォルダパス
#' @param subfolder サブフォルダ名（省略可）
#' @return 画像ファイルパスのベクトル
list_result_images <- function(result_dir, subfolder = NULL) {
  target_dir <- if (!is.null(subfolder)) {
    file.path(result_dir, subfolder)
  } else {
    result_dir
  }
  
  if (!dir.exists(target_dir)) {
    return(character())
  }
  
  list.files(target_dir, pattern = "\\.(png|jpg|jpeg)$", 
             full.names = TRUE, recursive = TRUE)
}

File: app/services/data_manager.py
# =============================================================================
# MSI Analysis Application - Data Manager
# データ管理モジュール
# =============================================================================

from pathlib import Path
from typing import Optional

import pandas as pd


def list_msi_files(data_folder: str) -> list[str]:
    """データフォルダ内のMSIファイル一覧を取得（拡張子なし）"""
    folder = Path(data_folder)
    if not folder.is_dir():
        return []
    return sorted([f.stem for f in folder.glob("*.txt")])


def list_tims_files(data_folder: str) -> list[str]:
    """データフォルダ内のTIMSファイル一覧を取得（拡張子なし）
    対応形式: .parquet, .pq, .csv, .tsv, .txt
    """
    folder = Path(data_folder)
    if not folder.is_dir():
        return []
    extensions = {".parquet", ".pq", ".csv", ".tsv", ".txt"}
    files = [f.stem for f in folder.iterdir() if f.suffix.lower() in extensions]
    return sorted(files)


def build_tims_input_paths(data_folder: str) -> list[str]:
    """データフォルダ内のTIMSファイルのフルパスリストを返す。
    Rスクリプトの INPUT_PATHS / ORIGINAL_INPUT_PATHS に対応。
    """
    folder = Path(data_folder)
    if not folder.is_dir():
        return []
    extensions = {".parquet", ".pq", ".csv", ".tsv", ".txt"}
    files = [str(f) for f in folder.iterdir() if f.suffix.lower() in extensions]
    return sorted(files)


def validate_msi_file(file_path: str) -> dict:
    """MSIファイルの妥当性チェック"""
    path = Path(file_path)
    if not path.exists():
        return {"valid": False, "message": "ファイルが見つかりません"}

    try:
        with open(path, "r", encoding="utf-8") as f:
            lines = [f.readline() for _ in range(5)]

        if len([l for l in lines if l]) < 5:
            return {"valid": False, "message": "ファイルの行数が不足しています (最低5行必要)"}

        # 5行目がタブ区切りデータかチェック
        data_fields = lines[4].strip().split("\t")
        if len(data_fields) < 4:
            return {"valid": False, "message": "データ形式が正しくありません"}

        return {"valid": True, "message": "OK"}
    except Exception as e:
        return {"valid": False, "message": f"読み込みエラー: {e}"}


def load_mrm_file(mrm_path: str) -> Optional[pd.DataFrame]:
    """MRMファイルの読み込み (xlsx/csv)"""
    if not mrm_path or not Path(mrm_path).exists():
        return None

    ext = Path(mrm_path).suffix.lower()

    try:
        if ext in (".xlsx", ".xls"):
            df = pd.read_excel(mrm_path)
        else:
            df = pd.read_csv(mrm_path)

        if df.empty:
            return None
        return df
    except Exception as e:
        print(f"MRMファイル読み込みエラー: {e}")
        return None


def list_result_folders(base_dir: str) -> list[dict]:
    """出力結果フォルダの一覧を取得"""
    base = Path(base_dir)
    if not base.is_dir():
        return []

    results = []
    for d in sorted(base.iterdir()):
        if not d.is_dir():
            continue

        # RDS_Filesサブフォルダがあれば解析結果フォルダと判定
        has_rds = (d / "RDS_Files").is_dir()
        has_outputs = any(d.glob("*.png")) or any(d.glob("*.csv"))

        if has_rds or has_outputs:
            import re
            date_match = re.search(r"\d{8}", d.name)
            if date_match:
                ds = date_match.group()
                date_str = f"{ds[:4]}/{ds[4:6]}/{ds[6:8]}"
            else:
                date_str = "Unknown"

            results.append({
                "name": d.name,
                "path": str(d),
                "date": date_str,
            })

    return results


def list_result_images(result_dir: str, subfolder: str = None) -> list[str]:
    """結果フォルダ内の画像ファイル一覧を取得"""
    target = Path(result_dir)
    if subfolder:
        target = target / subfolder

    if not target.is_dir():
        return []

    extensions = {".png", ".jpg", ".jpeg"}
    images = [
        str(f)
        for f in target.rglob("*")
        if f.suffix.lower() in extensions
    ]
    return sorted(images)
